### DISCUSSION TITLE
lammpsのstress tensor

### DISCUSSION THREAD

**Discussion by nim-hrkn** [Tomoki-YAMASHITA/CrySPY#39] [General]
lammpsのstress tensor
は
thermo_style  custom step etotal fmax vol press pxx pyy pzz pxy pxz pyz
でいいのでしょうか．ご存知ですか．

**Comment by Tomoki-YAMASHITA** [comment]
lammpsはあまり使っていないので私もよくわかりません．
とりあえず動かしている程度です．
ストレスの取り方もあまり知りません

**Comment by yw-fang** [comment]
For your information, generally I use the following flags for thermo_style
step temp pe ke press pxx pyy pzz pxy pxz pyz

**Reply by nim-hrkn** [reply]
Thank you for the information on the way to show the stress of LAMMPS!

**Comment by nim-hrkn** [comment]
'aiida.optimize' doesn't parse
```Python
'thermo_keywords': ["pxx", "pyy", "pzz", "pxy", "pxz", "pyz"]
```
in lammps parameter.

But, 
```Python
'output_variables': ["temp", "etotal", "pe", "ke" , 
                         "pxx", "pyy", "pzz", "pxy", "pxz", "pyz"],
```
in lammps parameter outputs
```
 'final_variables': {'ke': 0.0,
  'pe': 45.891137521116,
  'pxx': 414021.25432172,
  'pxy': -4.6129502655069e-10,
  'pxz': -5.8983989235005e-06,
  'pyy': 414021.25432172,
  'pyz': -5.8973854586867e-06,
  'pzz': 4011966.0768615,
  'temp': 0.0,
  'etotal': 45.891137521116},
```
in "results" Dict.

The units,
 'force_units': 'eV/Angstrom',
 'units_style': 'metal',
 'energy_units': 'eV',
 'parser_class': 'OptimizeParser',
 'stress_units': 'bars',
 'distance_units': 'Angstroms',
are kindly written in attributes. In addition, forces are parsed already in aiida-lammps. 

LAQA uses only the last stress, force and total energy values. Then, LAQA features can be added without changing or adding aiida-lammps parser.

**Reply by nim-hrkn** [reply]
When the 'thermo_keywords' and 'output_variables' are added. aiida.optimize generates the input file, that is
```
units          metal
boundary        p p p
box tilt large
atom_style      atomic
read_data       input.data
pair_style      tersoff
pair_coeff      * * potential.pot Ga N
fix             int all box/relax aniso 0.0 vmax 0.005
compute         stpa all stress/atom NULL
compute         stgb all reduce sum c_stpa[1] c_stpa[2] c_stpa[3] c_stpa[4] c_stpa[5] c_stpa[6]
variable        stress_pr equal -(c_stgb[1]+c_stgb[2]+c_stgb[3])/(3*vol)
thermo_style custom step temp press etotal v_stress_pr pxx pyy pzz pxy pxz pyz
dump            aiida all custom 1 trajectory.lammpstrj element x y z  fx fy fz c_stpa[1] c_stpa[2] c_stpa[3] c_stpa[4] c_stpa[5] c_stpa[6]
dump_modify     aiida format line "%4s %16.10f %16.10f %16.10f %16.10f %16.10f %16.10f %16.10f %16.10f %16.10f %16.10f %16.10f %16.10f"
dump_modify     aiida sort id
dump_modify     aiida element Ga N
min_style       cg
minimize        1e-05 0.001 5 10000
variable temp equal temp
print "final_variable: temp = ${temp}"
variable etotal equal etotal
print "final_variable: etotal = ${etotal}"
variable pe equal pe
print "final_variable: pe = ${pe}"
variable ke equal ke
print "final_variable: ke = ${ke}"
variable pxx equal pxx
print "final_variable: pxx = ${pxx}"
variable pyy equal pyy
print "final_variable: pyy = ${pyy}"
variable pzz equal pzz
print "final_variable: pzz = ${pzz}"
variable pxy equal pxy
print "final_variable: pxy = ${pxy}"
variable pxz equal pxz
print "final_variable: pxz = ${pxz}"
variable pyz equal pyz
print "final_variable: pyz = ${pyz}"
variable final_energy equal etotal
print "final_energy: ${final_energy}"
print "END_OF_COMP"
``
Note that stress/atom is calculated, but isn't parsed.

**Reply by yw-fang** [reply]
> 'aiida.optimize' doesn't parse
> 
> ```python
> 'thermo_keywords': ["pxx", "pyy", "pzz", "pxy", "pxz", "pyz"]
> ```
> 


Thank you for proceeding with it. It seems that I missed something. I'm wondering where is the "aiida.optimize"?
I didn't find it in the default branches of aiida-core, aiida-lammps or aiida-cryspy.

**Reply by nim-hrkn** [reply]
Excuse me for writing the wrong term.
wrong: aiida.optimize
correct: lammps.optimize
It is one of aiida-lammps plugins, as you know.
https://github.com/aiidaplugins/aiida-lammps

**Reply by yw-fang** [reply]
OK. Thank you for the fix.

**Comment by nim-hrkn** [comment]
It looks that ctrl_collect_laqa() doesn't use the stress.
The forces are zero if the unoptimized cell is a factor times the optimized cell. 
Is LAQA designed for the case without cell optimization?

**Reply by Tomoki-YAMASHITA** [reply]
Even if you perform cell optimization, LAQA currently uses only energy and force without stress.
The current score is based on Terayama 2018(https://www.nature.com/articles/s41524-018-0090-y).

The improved score (Yamashita 2022; https://www.tandfonline.com/doi/full/10.1080/27660400.2022.2059335) is not implemented yet.

**Reply by nim-hrkn** [reply]
THank you for the reply. 
The weight of the force term is rin.weight_laqa. Which name is used for the weight of the stress term? I use the same name.

**Reply by Tomoki-YAMASHITA** [reply]
Currently not implemented, but I will change the name (next version?):
- weight_laqa --> wf_laqa
- (new) --> ws_laqa

**Reply by nim-hrkn** [reply]
Thank you for the explanation and letting me know the new names. 
The related original parts, e.g. ctrl_collect_{RS|EA|BO}, weren't used in aiida-crspy because the physical properties are obtained not from the files, but from the output ports of aiida-plugins. Therefore, I rewrote the related code of RS, EA and BO in aiida-crspy.  So will be LAQA. Probably I understood how to do it from the paper.
